ctapipe/image/muon/processor.py

```diff
@@ -0,0 +1,265 @@
+        ),
+    ).tag(config=True)
+
+    pedestal = FloatTelescopeParameter(
+        help="Pedestal noise rms", default_value=1.1
+    ).tag(config=True)
+
+    def __init__(self, subarray, **kwargs):
+        super().__init__(subarray, **kwargs)
+        self.dl1_query = DL1Query(parent=self)
+        self.ring_query = RingQuery(parent=self)
+        self.geometries = {
+            tel_id: tel.camera.geometry.transform_to(TelescopeFrame())
+            for tel_id, tel in self.subarray.tel.items()
+        }
+        self.fov_radius = {
+            tel_id: geometry.guess_radius()
+            for tel_id, geometry in self.geometries.items()
+        }
+
+        self.ring_fitter = MuonRingFitter(parent=self)
+
+        self.intensity_fitter = MuonIntensityFitter(subarray=subarray, parent=self)
+
+    def __call__(self, event: ArrayEventContainer):
+        for tel_id in event.dl1.tel:
+            self._process_telescope_event(event, tel_id)
+
+    def _process_telescope_event(self, event, tel_id):
+        """
+        Extract and process a ring from a single image.
+
+        Parameters
+        ----------
+        event: ArrayEventContainer
+            Collection of all event information
+        tel_id: int
+            Telescope ID of the instrument that has measured the image
+        """
+        event_index = event.index
+        event_id = event_index.event_id
+
+        if self.subarray.tel[tel_id].optics.n_mirrors != 1:
+            self.log.warning(
+                f"Skipping non-single mirror telescope {tel_id},"
+                " muon-fitting for dual-mirror telescope is currently"
+                " not supported. Exclude dual mirror telescopes via setting"
+                " 'EventSource.allowed_tels'."
+            )
+            event.muon.tel[tel_id] = MuonCameraContainer()
+            return
+
+        self.log.debug(f"Processing event {event_id}, telescope {tel_id}")
+        dl1 = event.dl1.tel[tel_id]
+        image = dl1.image
+        mask = dl1.image_mask
+        if mask is None:
+            mask = image > 0
+
+        checks = self.dl1_query(p=dl1.parameters)
+
+        if not all(checks):
+            event.muon.tel[tel_id] = MuonCameraContainer()
+            return
+
+        geometry = self.geometries[tel_id]
+        fov_lon = geometry.pix_x
+        fov_lat = geometry.pix_y
+
+        # iterative ring fit.
+        # First use cleaning pixels, then only pixels close to the ring
+        # three iterations seems to be enough for most rings
+        for _ in range(3):
+            ring = self.ring_fitter(fov_lon, fov_lat, image, mask)
+            dist = np.sqrt(
+                (fov_lon - ring.center_fov_lon) ** 2
+                + (fov_lat - ring.center_fov_lat) ** 2
+            )
+            mask = np.abs(dist - ring.radius) / ring.radius < 0.4
+
+        parameters = self._calculate_muon_parameters(
+            tel_id, image, dl1.image_mask, ring
+        )
```

Comment:
The result from the ring fitter is used here but not the `mask`?

Reply:
Yes, these statistics partially compare the cleaned image to the fitted ring, see e.g. `ctapipe.image.muon.features.intensity_ratio`. There might indeed be a better way by comparing the cleaned image and the fitted ring mask but I'd rather keep this the way it was done before and then create a new PR for it then changing it in this PR too.

Reply:
Yes, see above comment about having this PR only do the refactoring and not do changes in behavior.